### PR TITLE
Add Forced Low Memory Mode feature

### PR DESCRIPTION
**Pull request by mattwar**
Issue #11354 

This adds an option that lets Roslyn developers put their VS process into a low memory mode. This will simulate a VS install with more constrained address space.

@mattgertz @jasonmalinowski @jared @Pilchie please review

### CODE REVIEW

File: src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs [hunk 1]
```diff
@@ -0,0 +1,141 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Roslyn.VisualStudio.DiagnosticsWindow.OptionsPages
```
Comment (jasonmalinowski) [inline src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs]: Just FYI, I'm not sure how widely consumed this project is, or whether anybody ever installs it. I was considering deleting it...but this is a fine use. You might have a (small) uphill battle to get people to install it again.

File: src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                            block[i] = 0xFF;
+                        }
+
+                        _blocks.Add(block);
+
+                        // don't hog the thread
+                        await Task.Yield();
+                    }
+                }
+                catch (Exception)
```
Comment (jasonmalinowski) [inline src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs]: Can we be a bit more precise here?

File: src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+                        // initialize block bits (so they memory actually gets allocated.. silly runtime!)
+                        for (int i = 0; i < BlockSize; i++)
+                        {
+                            block[i] = 0xFF;
+                        }
+
+                        _blocks.Add(block);
+
+                        // don't hog the thread
+                        await Task.Yield();
```
Comment (jasonmalinowski) [inline src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs]: Consider just starting your task with TaskCreationOptions.LongRunning which will just give you your own thread.

Comment (jasonmalinowski) [inline src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs]: Oh I see, you're also polling later...hmm...

File: src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                    try
+                    {
+                        // access all block cells
+                        for (var b = 0; b < _blocks.Count && !_cancelled; b++)
+                        {
+                            var block = _blocks[b];
+
+                            for (int i = 0; i < block.Length; i++)
+                            {
+                                var tmp = block[i]; // read bytes from block
```
Comment (jasonmalinowski) [inline src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs]: Is the compiler or runtime entitled to optimize this away? @jaredpar?

File: src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        public static readonly ForceLowMemoryMode Instance = new ForceLowMemoryMode();
+
+        private ForceLowMemoryMode()
+        {
+        }
+
+        public int Size
+        {
+            get { return _size; }
+            set { _size = value; }
```
Comment (jasonmalinowski) [inline src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs]: Does changing this need to restart the hogger?

**Comment by jasonmalinowski** [comment]
:+1:

File: src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs (hunk 5 above)
Comment (mattwar) [inline src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs]: No, you have to disable the low memory mode, change the value and re-enable to get it to hog a different amount of megabytes.

File: src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs (hunk 1 above)
Comment (mattwar) [inline src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs]: I think it comes along with install dogfood.

File: src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs (hunk 2 above)
Comment (mattwar) [inline src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs]: Do you mean with the exception type?   This is really just meant to capture out-of-memory exceptions.

File: src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs (hunk 5 above)
Comment (mattwar) [inline src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs]: probably should be auto-prop... oh well.

File: src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs (hunk 4 above)
Comment (jaredpar) [inline src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs]: @jasonmalinowski possibly but we don't.

File: src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+                    if (hogger != null)
+                    {
+                        _hogger = null;
+                        hogger.Cancel();
+                    }
+                }
+            }
+        }
+
+        class MemoryHogger
```
Comment (jaredpar) [inline src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs]: public

File: src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+            private async Task PopulateAndMonitorWorkerAsync(int size)
+            {
+                try
+                {
+                    for (int n = 0; n < size && !_cancelled; n++)
+                    {
+                        var block = new byte[BlockSize];
+
+                        // initialize block bits (so they memory actually gets allocated.. silly runtime!)
```
Comment (bkoelman) [inline src/VisualStudio/VisualStudioDiagnosticsToolWindow/OptionPages/ForceLowMemoryMode.cs]: they -> _the_

**Pull request merged**